owlkettle/adw.nim

```diff
@@ -867,6 +870,157 @@ when AdwVersion >= (1, 2) or defined(owlkettleDocs):
   
   export AboutWindow
 
+## Adw.Toast
+type Toast* = ref object
+  title*: string
+  customTitle*: Widget
+  buttonLabel*: string
+  priority*: ToastPriority
+  timeout*: int
+  dismissalHandler: proc(toast: Toast)
+  clickedHandler: proc()
+  useMarkup: bool
+
+proc newToast*(
+  title: string,
+  buttonLabel: string = "", 
+  priority: ToastPriority = ToastPriorityNormal, 
+  dismissalHandler: proc(toast: Toast) = nil, 
+  clickedHandler: proc() = nil,
+  timeout: int = 5, 
+  useMarkup: bool = false,
+  customTitle: Widget = nil.Widget
+): Toast =
+  result = Toast(
+    title: title, 
+    buttonLabel: buttonLabel,
+    priority: priority,
+    timeout: timeout,
+    dismissalHandler: dismissalHandler
+  )
+  
+  when AdwVersion >= (1, 2):
+    result.clickedHandler = clickedHandler
+    result.customTitle = customTitle
+  else:
+    let isUsingCustomTitle = not customTitle.isNil()
+    if isUsingCustomTitle:
+      raise newException(LibraryError, "The customTitle field on Toast is not available when compiling for Adwaita versions below 1.2. Compile for Adwaita version 1.2 or higher with -d:adwminor=2 to enable it")
+
+    let isUsingClickedHandler = not clickedHandler.isNil()
+    if isUsingClickedHandler:
+      raise newException(LibraryError, "The clickedHandler field on Toast is not available when compiling for Adwaita versions below 1.2. Compile for Adwaita version 1.2 or higher with -d:adwminor=2 to enable it")
+      
+  when AdwVersion >= (1, 4):
+    result.useMarkup = useMarkup
+  else:
+    let isUsingUseMarkup = useMarkup == true
+    if isUsingUseMarkup:
+      raise newException(LibraryError, "The useMarkup on Toast field is not available when compiling for Adwaita versions below 1.4. Compile for Adwaita version 1.4 or higher with -d:adwminor=4 to enable it")
+
+proc toOwl(adwToast: AdwToast): Toast =
+  when AdwVersion >= (1, 4):
+    let useMarkup = adw_toast_get_use_markup(adwToast).bool
+  else:
+    let useMarkup = false
+  
+  result = newToast(
+    title = $adw_toast_get_title(adwToast),
+    buttonLabel = $adw_toast_get_button_label(adwToast),
+    priority = adw_toast_get_priority(adwToast),
+    timeout = adw_toast_get_timeout(adwToast).int,
+    useMarkup = useMarkup
+  )
+  
+proc toGtk(toast: Toast): AdwToast =
+  result = adw_toast_new(toast.title.cstring)
+  if toast.buttonLabel != "":
+    adw_toast_set_button_label(result, toast.buttonLabel.cstring)
+  adw_toast_set_priority(result, toast.priority)
+  let timeout: cuint = cuint(toast.timeout)
+  adw_toast_set_timeout(result, timeout)
+  
+  # Set Dismissal Handler
+  if not toast.dismissalHandler.isNil():
```

Comment:
I wonder whether I should've moved the contents of those if-blocks into their own procs just to make this a better separation of concerns in terms of the procs

Reply:
I did that now for simplification purposes